agent/consul/state/catalog.go

```diff
@@ -2363,79 +2413,125 @@ func checkSessionsTxn(tx *memdb.Txn, hc *structs.HealthCheck) ([]*sessionCheck,
 	return sessions, nil
 }
 
-// updateGatewayService associates services with gateways as specified in a terminating-gateway config entry
-func (s *Store) updateTerminatingGatewayServices(tx *memdb.Txn, idx uint64, conf structs.ConfigEntry, entMeta *structs.EnterpriseMeta) error {
-	entry, ok := conf.(*structs.TerminatingGatewayConfigEntry)
-	if !ok {
-		return fmt.Errorf("unexpected config entry type: %T", conf)
-	}
+// updateGatewayServices associates services with gateways as specified in a gateway config entry
+func (s *Store) updateGatewayServices(tx *memdb.Txn, idx uint64, conf structs.ConfigEntry, entMeta *structs.EnterpriseMeta) error {
+	var gatewayServices structs.GatewayServices
+	var err error
 
-	// Check if service list matches the last known list for the config entry, if it does, skip the update
-	_, c, err := s.configEntryTxn(tx, nil, conf.GetKind(), conf.GetName(), entMeta)
-	if err != nil {
-		return fmt.Errorf("failed to get config entry: %v", err)
+	gatewayID := structs.NewServiceID(conf.GetName(), conf.GetEnterpriseMeta())
+	switch conf.GetKind() {
+	case structs.IngressGateway:
+		gatewayServices, err = s.ingressConfigGatewayServices(tx, gatewayID, conf, entMeta)
+	case structs.TerminatingGateway:
+		gatewayServices, err = s.terminatingConfigGatewayServices(tx, gatewayID, conf, entMeta)
```

Comment:
Do you want an explicit `default` branch here to return a "wrong kind" error, or is the fallthrough to `return nil` ok?

Reply:
Added an error check here so that we could remove the `len() == 0` check below